<launch>
<!-- Top-Level launch file for bundled launching /-->
<!-- Launcher for the uav_ftc package -->

	<!-- Available arguments -->
	<arg name="manualMode" default="false"/>
	<arg name="autoMode" default="false"/>
	<arg name="uav_name" default="default_mot_0"/>
	<arg name="log" default="true"/>
    <arg name="fe_name" default="default_mot_0"/>

    <!-- Launch uav_ftc-related nodes -->
    <group ns="$(arg uav_name)">
        <!-- Load controller parameters -->
        <param name="uav_name" value="$(arg uav_name)"/>
        <rosparam>
            ctrlMode: 2
            pMin: -1.5
            qMin: -1.5
            rMin: -0.25
            pMax: 1.5
            qMax: 1.5
            rMax: 0.25
            referenceMin: [10.0, -0.52, -0.5]
            referenceMax: [20.0,  0.52,  0.5]
            wp: 10
            wq: 10
            wr: 10
            wda: 100
            wde: 100
            wdr: 10
            wnp: 1000
            wnq: 1000
            wnr: 1000
            wgamma: 1000.0
            wpsidot: 10000.0
            walpha: 0.01
            wbeta: 100
            wdp: 10.0
            wdq: 50.0
            wdr: 100.0
            wphi: 0.01
            wngamma: 10000.0
            wnphi: 100.0
        </rosparam>

        <!--  -->
        <!-- Launch data sources -->
        <!--  -->
        <include file="$(find uav_ftc)/launch/data_sources.launch" />

        <!-- Launch static FE publisher -->
        <node pkg="uav_ftc" name="fe_publisher" type="fe_publisher.py" output="screen">
            <param name="fe" value="$(arg fe_name)"/>
        </node>

        <!-- Raise rate controller -->
        <!-- <node pkg="uav_ftc" name="rate_controller" type="rate_controller" respawn="true" output="screen">
            <param name="~rate" value="50"/>
            <remap from="~wp" to="wp"/>
            <remap from="~wq" to="wq"/>
            <remap from="~wr" to="wr"/>
            <remap from="~wda" to="wda"/>
            <remap from="~wde" to="wde"/>
            <remap from="~wdr" to="wdr"/>
            <remap from="~wnp" to="wnp"/>
            <remap from="~wnq" to="wnq"/>
            <remap from="~wnr" to="wnr"/>
        </node> -->
        <node pkg="uav_ftc" name="rate_controller" type="rate_controller_pid" respawn="true" output="screen">
            <param name="~rate" value="50"/>
            <param name="~kp_p" value="0.6"/>
            <param name="~ki_p" value="0.06"/>
            <param name="~kd_p" value="0"/>
            <param name="~kp_q" value="0.6"/>
            <param name="~ki_q" value="0.06"/>
            <param name="~kd_q" value="0"/>
            <param name="~kp_r" value="1.2"/>
            <param name="~ki_r" value="0.12"/>
            <param name="~kd_r" value="0"/>
        </node>

        <!-- Raise velocity controller -->
        <node pkg="uav_ftc" name="trajectory_controller" type="trajectory_controller_no_va" respawn="true" output="screen">
            <param name="~rate" value="5"/>
            <param name="~kp_va" value="0.1"/>
            <param name="~ki_va" value="0.05"/>
            <param name="~kd_va" value="0"/>
            <remap from="~pMin" to="pMin"/>
            <remap from="~qMin" to="qMin"/>
            <remap from="~rMin" to="rMin"/>
            <remap from="~pMax" to="pMax"/>
            <remap from="~qMax" to="qMax"/>
            <remap from="~rMax" to="rMax"/>
            <remap from="~wgamma" to="wgamma"/>
            <remap from="~wpsidot" to="wpsidot"/>
            <remap from="~walpha" to="walpha"/>
            <remap from="~wbeta" to="wbeta"/>
            <remap from="~wdp" to="wdp"/>
            <remap from="~wdq" to="wdq"/>
            <remap from="~wdr" to="wdr"/>
            <remap from="~wphi" to="wphi"/>
            <remap from="~wngamma" to="wngamma"/>
            <remap from="~wnphi" to="wnphi"/>
        </node>

        <!-- Raise input aggreggator node -->
        <node pkg="uav_ftc" name="input_aggregator" type="input_aggregator" respawn="true" output="screen">
            <param name="~rate" value="100"/>
        </node>

        <!-- Raise reference generator node -->
        <node pkg="uav_ftc" name="reference_generator" type="reference_generator" respawn="true" output="screen">
            <remap from="~ctrlMode" to="ctrlMode"/>
            <remap from="~referenceMin" to="referenceMin"/>
            <remap from="~referenceMax" to="referenceMax"/>
            <remap from="refCmds" to="refTrajectory"/>
        </node>

        <!-- Raise rosflight topic converted node -->
        <node pkg="uav_ftc" name="rosflight_publisher" type="rosflight_publisher" respawn="true" output="screen">
            <param name="xscale" value="1.0"/>
            <param name="yscale" value="1.0"/>
            <param name="zscale" value="1.0"/>
            <param name="fscale" value="0.0"/>
        </node>

        <!--  -->
        <!-- Launch logger -->
        <!--  -->
        <include if="$(arg log)" file="$(find uav_ftc)/launch/logger.launch" />

    </group>

</launch>